src/celttf/truetypefont.cpp

```diff
@@ -646,69 +708,89 @@ TextureFont* TextureFont::load(const Renderer *r, const fs::path &path, int inde
     if (!FT_IS_SCALABLE(face))
     {
         GetLogger()->error("Font is not scalable: {}\n", path);
+        FT_Done_Face(face);
         return nullptr;
     }
 
     if (FT_Set_Char_Size(face, 0, size << 6, dpi, dpi) != 0)
     {
         GetLogger()->error("Could not set font size {}\n", size);
+        FT_Done_Face(face);
         return nullptr;
     }
 
-    auto* font = new TextureFont(r);
-    font->impl->m_face = face;
-
-    if (!font->impl->buildAtlas())
-        return nullptr;
-
-    font->setMaxAscent(face->size->metrics.ascender >> 6);
-    font->setMaxDescent(-face->size->metrics.descender >> 6);
-
-    return font;
+    return face;
 }
 
 // temporary while no fontconfig support
-static fs::path ParseFontName(const fs::path &filename, int &collectionIndex, int &size)
+fs::path
+ParseFontName(const fs::path &filename, int &index, int &size)
 {
     // Format with font path/collection index(if any)/font size(if any)
     auto fn = filename.string();
-    auto pos = fn.rfind(',');
-    if (pos != string::npos)
+    if (auto ps = fn.rfind(','); ps != std::string::npos)
     {
-        size = (int) stof(fn.substr(pos + 1));
-        auto rest = fn.substr(0, pos);
-
-        pos = rest.rfind(',');
-        if (pos != string::npos)
+        if (auto [_, ec] = from_chars(&fn[ps + 1], &fn[fn.size()], size); ec == std::errc())
         {
-            collectionIndex = stof(rest.substr(pos + 1));
-            return rest.substr(0, pos);
+            if (auto pi = fn.rfind(',', ps - 1); pi != std::string::npos)
+            {
+                if (auto [_, ec] = from_chars(&fn[pi + 1], &fn[pi], index); ec == std::errc())
+                    return fn.substr(0, pi);
+            }
+            return fn.substr(0, ps);
         }
-        else
-        {
-            return rest;
-        }
-    }
-    else
-    {
-        size = 12;
-        return filename;
     }
+    return filename;
 }
 
-std::shared_ptr<TextureFont> LoadTextureFont(const Renderer *r, const fs::path &filename, int index, int size)
+} // namespace
+
+using FontCache = std::map<fs::path, std::weak_ptr<TextureFont>>;
+
+std::shared_ptr<TextureFont>
+LoadTextureFont(const Renderer *r, const fs::path &filename, int index, int size)
 {
-    if (ft == nullptr)
+    // Init FreeType library
+    static FT_Library ftlib = nullptr;
+    if (ftlib == nullptr && FT_Init_FreeType(&ftlib) != 0)
     {
-        if (FT_Init_FreeType(&ft))
+        GetLogger()->error("Could not init freetype library\n");
+        return nullptr;
+    }
+
+    // Init FontCache
+    static FontCache *fontCache = nullptr;
+    if (fontCache == nullptr)
+        fontCache = new FontCache;
+
+    // Lookup for an existing cached font
+    std::weak_ptr<TextureFont> &font = (*fontCache)[filename];
+    if (font.expired())
```

Comment:
I think it's better to use the result of `lock` for the test, rather than testing `expired` and doing the `lock` later

```C++
std::shared_ptr<TextureFont> ret = font.lock();
if (ret == nullptr)
{
    // do the creation
}
return ret;

Reply:
probably such way is safer